File: 2019/knowit/23/23.cpp
#include <iostream>

int digit_sum(int n) {
	int sum = 0;
	while (n > 0) {
		sum += n % 10;
		n = n / 10;
	}

	return sum;
}

bool is_prime(int n) {
        if (n == 1) return false;
        if (n < 4) return true;
        if (n % 2 == 0) return false;
        if (n % 3 == 0) return false;

        int i = 5;
        int w = 2;

        while (i * i <= n) {
                if (n % i == 0) return false;
                i += w;
                w = 6 - w;
        }
        return true;
}

bool is_harshad(int n) {
	int ds = digit_sum(n);
	return is_prime(ds) && n % ds == 0;
}

int main(int argc, char** argv) {
	int limit = 98765432;

	int count = 0;
	for (int n = 1; n <= limit; n++) {
		if (is_harshad( n)) {
			count++;
		}
	}

	std::cout << count << std::endl;
}

File: 2019/knowit/23/23.hs
import qualified Data.Set as S

-- Note: Digits are in reverse order
type Digits = [Int]

limit :: Int
limit = 98765432

primes :: S.Set Int
primes = S.fromAscList . takeWhile (<maxPrime) $ 2 : 3 : 5 : primes'
    where isPrime (p:ps) n = p*p > n || n `rem` p /= 0 && isPrime ps n
          primes' = 7 : filter (isPrime primes') (scanl (+) 11 $ cycle [2,4,2,4,6,2,6,4])
          maxPrime = (*9) . length . toDigits $ limit

toDigits :: Int -> Digits
toDigits 0 = []
toDigits n = n `mod` 10 : toDigits (n `div` 10)

digitSum :: Int -> Int
digitSum = sum . toDigits

isHarshad :: Int -> Bool
isHarshad n = let ds = digitSum n
              in  n `mod` ds == 0 && S.member ds primes

main =
    print
        . length
        . filter isHarshad
        $ [1..limit]
